Make polluting global scope optional (alternative)

aiham opened the pull request:
Alternative solution to #547

aiham commented:
I signed it!

aiham commented:
Thanks. I'll start working on resolving the conflicts.

fippo commented on Gruntfile.js:
```diff
@@ -46,6 +46,12 @@ module.exports = function(grunt) {
+      // with a custom window object.
+      adapterFactory: {
+        src: ['./src/js/adapter_factory.js'],
+        dest: './out/adapter_factory.js'
```
is it necessary to provide this? If you want the factory you'd need to require/import the module, i wouldn't want to make our life harder with this ;-)

aiham replied:
Good point. We can access it with `const adapterFactory = require('webrtc-adapter/src/js/adapter_factory.js');` we just have to be careful in case the file structure changes in the future.

I'll remove this.

aiham commented:
@fippo I've updated the branch. There was something I wanted to change that might be a little controversial so I split it into its own commit: 509a50fc3d64842b7d3f0b3628b9e626f5ebbaa1

Removes `utils.browserDetails`. Without this commit the browser detection always runs agains the global window. With this commit almost all unit tests are completely isolated from each other.

WDYT?

aiham commented:
@fippo @KaptenJansson I'm going to wait on getting the final green light on these 2 commits before resolving the conflicts again otherwise I'll probably end up repeating myself a bunch of times. Just let me know once you're happy with it, I'll resolve the conflicts and we can get it merged in before anything else causes another conflict. Thanks!

fippo commented:
:+1: -- even if the only reason was to make testing more isolated!

fippo commented:
the last commit will conflict with #550 -- but i can rebase,

aiham commented:
@fippo Oh you're converting them to unit tests, that's great. I just spent some time trying to fake the Firefox version in the integrations. I think going with unit tests is the better approach. It would be great if you could rebase. Should I disable these tests so my PR passes?

fippo commented:
Might be easier if you revert your last commit and then merge mine. Then its no longer your problem ;-)

KaptenJansson commented:
This looks great! LGTM once the test passes, will hold of reviewing/merging other PR's until this is in so you can solve all conflicts.

aiham commented:
@fippo Done!

fippo commented:
argh... so much for "that plan is easier".

consenting to inclusion obviously.

aiham commented:
I think the chrome beta test will pass if restarted. It timed out.

fippo commented:
\o/

@KaptenJansson so lets merge this and i'll try to re-land #527 with a native implementation for edge

KaptenJansson commented:
LGTM! Thanks for doing this!

fippo commented:
I wonder if we should remove https://github.com/fippo/eslint-config-webrtc/blob/master/index.js now. That would prevent us from accidentally using those objects, right?

fippo commented:
pity, it looks like eslint knows most of these things nowadays.

aiham commented:
@fippo @KaptenJansson :tada: Thanks for helping to get this through!
